src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -26,25 +43,83 @@
-  */
+   * Search and returns a list of first 10 recipes with the title matching the given parameter(searchterm).
+   * The index returns a list of documents in the ascending order by title
+   */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response)
     throws ServletException, IOException {
-    request.setAttribute("messages", request.getParameter("searchterm"));
-
-    ArrayList<Recipe> recipes = new ArrayList<>();
-    for (int i = 0; i < 5; ++i) {
-      Recipe recipe = new Recipe();
-      recipe.setId(i);
-      recipe.setName("Cake");
-      recipe.setDescription("Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.");
-      recipes.add(recipe);
+    int error = 0;
+    List<Recipe> recipesList = new ArrayList<>();
+    String titleToMatch = request.getParameter("searchterm");
+    Index index = getIndex();
+    Cursor responseCursor = Cursor.newBuilder().build();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    QueryOptions options = QueryOptions
+      .newBuilder()
+      .setLimit(10)
+      .setFieldsToSnippet("title")
+      .setCursor(responseCursor)
+      .setSortOptions(
+        SortOptions
+          .newBuilder()
+          .addSortExpression(
+            SortExpression
+              .newBuilder()
+              .setExpression("title")
+              .setDirection(SortExpression.SortDirection.ASCENDING)
+              .setDefaultValue("")
+          )
+      )
+      .build();
+
+    Query query = Query
+      .newBuilder()
+      .setOptions(options)
+      .build(titleToMatch.toLowerCase());
+
+    Results<ScoredDocument> results = index.search(query);
+
+    for (ScoredDocument entity : results) {
+      try {
+        String id = entity.getId();
+        Entity recipeEntity = datastore.get(
+          KeyFactory.createKey("Recipe", Long.parseLong(id))
+        );
+
+        String name = (String) recipeEntity.getProperty("title");
+
+        Recipe recipe = new Recipe();
+        String description =
+          "Lorem quam dolor dapibus ante, sit amet pellentesque turpis lacus eu ipsum. Duis quis mi ut tortor interdum efficitur quis at mi. Pellentesque quis mauris vel ligula commodo scelerisque. In vulputate quam nisl, vel sagittis ipsum molestie quis. Suspendisse quis ipsum a sem aliquam euismod mattis sed metus.";
+
+        recipe.setId(Long.parseLong(id));
+        recipe.setName(name);
+        recipe.setImage((String) recipeEntity.getProperty("imgURL"));
+        recipe.setDescription(description);
```

Comment:
Please modularize this into a function called "buildRecipe"

Reply:
Also

"String name = (String) recipeEntity.getProperty("title");"

Should be below the initialization of Recipe.